src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -350,6 +364,122 @@ public override bool Exists(DataSourceObjectMetadata objectMetadata)
+
+        /// <inheritdoc/>
+        public override LanguageServices.Contracts.CompletionItem[] GetAutoCompleteSuggestions(ScriptDocumentInfo scriptDocumentInfo, Position textPosition, bool throwOnError = false){
+            var kustoCodeService = new KustoCodeService(scriptDocumentInfo.Contents, SchemaState);
+            var script = CodeScript.From(scriptDocumentInfo.Contents, SchemaState);
+            script.TryGetTextPosition(textPosition.Line + 1, textPosition.Character, out int position);     // Gets the actual offset based on line and local offset
+            
+            var completion = kustoCodeService.GetCompletionItems(position);
+            scriptDocumentInfo.ScriptParseInfo.CurrentSuggestions = completion.Items;         // this is declaration item so removed for now, but keep the info when api gets updated
+
+            List<LanguageServices.Contracts.CompletionItem> completions = new List<LanguageServices.Contracts.CompletionItem>();
+            foreach (var autoCompleteItem in completion.Items)
+            {
+                var label = autoCompleteItem.DisplayText;
+                completions.Add(AutoCompleteHelper.CreateCompletionItem(label, label + " keyword", label, KustoIntellisenseHelper.CreateCompletionItemKind(autoCompleteItem.Kind), scriptDocumentInfo.StartLine, scriptDocumentInfo.StartColumn, textPosition.Character));
+            }
+
+            return completions.ToArray();
+        }
+
+        /// <inheritdoc/>
+        public override Hover GetHoverHelp(ScriptDocumentInfo scriptDocumentInfo, Position textPosition, bool throwOnError = false){
+            var kustoCodeService = new KustoCodeService(scriptDocumentInfo.Contents, SchemaState);
+            var script = CodeScript.From(scriptDocumentInfo.Contents, SchemaState);
+            script.TryGetTextPosition(textPosition.Line + 1, textPosition.Character, out int position);
+
+            var quickInfo = kustoCodeService.GetQuickInfo(position);
+
+            return AutoCompleteHelper.ConvertQuickInfoToHover(
+                                        quickInfo.Text,
+                                        "kusto",
+                                        scriptDocumentInfo.StartLine,
+                                        scriptDocumentInfo.StartColumn,
+                                        textPosition.Character);
+
+        }
+
+        /// <inheritdoc/>
+        public override DefinitionResult GetDefinition(string queryText, int index, int startLine, int startColumn, bool throwOnError = false){
+            var abc = KustoCode.ParseAndAnalyze(queryText, SchemaState);        //TODOKusto: API wasnt working properly, need to check that part.
+            var kustoCodeService = new KustoCodeService(abc);
+            //var kustoCodeService = new KustoCodeService(queryText, globals);
+            var relatedInfo = kustoCodeService.GetRelatedElements(index);
+
+            if (relatedInfo != null && relatedInfo.Elements.Count > 1)
+            {
+
```

Comment:
To be implemented?

Reply:
yes, this one I need to implement later.